tracing-core/src/metadata.rs

```diff
@@ -91,18 +93,149 @@ pub struct Metadata<'a> {
 pub struct Kind(KindInner);
 
 /// Describes the level of verbosity of a span or event.
+///
+/// # Comparing Levels
+///
+/// `Level` implements the [`PartialOrd`] and [`Ord`] traits, allowing two
+/// `Level`s to be compared to determine which is considered more or less
+/// verbose. Levels which are more verbose are considered "greater than" levels
+/// which are less verbose, with [`Level::ERROR`] considered the lowest, and
+/// [`Level::TRACE`] considered the highest.
+///
+/// For example:
+/// ```
+/// use tracing_core::Level;
+///
+/// assert!(Level::TRACE > Level::DEBUG);
+/// assert!(Level::ERROR < Level::WARN);
+/// assert!(Level::INFO <= Level::DEBUG);
+/// assert_eq!(Level::TRACE, Level::TRACE);
+/// ```
+///
+/// # Filtering
+///
+/// Typically, `Level`s are used to implement *filtering* that determines which
+/// spans and events are enabled. Depending on the use case, more or less
+/// verbose diagnostics may be desired &mdash; for example, when running in
+/// development, [`DEBUG`]-level traces may be enabled by default, and when
+/// running in production, we might choose to enable only [`INFO`]-level and
+/// lower. Libraries may include very verbose diagnostics at the [`DEBUG`]
+/// and/or [`TRACE`] levels, which applications using those libraries might
+/// typically choose to ignore. However, when debugging an issue involving those
+/// libraries, it may be useful to temporarily enable those traces.
+///
+/// The [`LevelFilter`] type is provided to enable filtering traces by
+/// verbosity. `Level`s can be compared against [`LevelFilter`]s, and
+/// [`LevelFilter`] has a variant for each `Level`, which compares analogously
+/// to that level. In addition, [`LevelFilter`] adds a [`LevelFilter::OFF`]
+/// variant, which is considered "less verbose" than every other `Level. This is
+/// intended to allow filters to completely disable tracing in a particular context.
+///
+/// For example:
+/// ```
+/// use tracing_core::{Level, LevelFilter};
+///
+/// assert!(LevelFilter::OFF < Level::TRACE);
+/// assert!(LevelFilter::TRACE > Level::DEBUG);
+/// assert!(LevelFilter::ERROR < Level::WARN);
+/// assert!(LevelFilter::INFO <= Level::DEBUG);
+/// assert!(LevelFilter::INFO >= Level::INFO);
+/// ```
+///
+/// As a simple example of how a [collector] could implement filtering, a
+/// [`LevelFilter`] is provided when the collector is constructed. Then, when a
+/// span or event is recorded, the [`Collect::enabled`] method would compare the
+/// span or event's `Level` against the configured [`LevelFilter`]. The optional
+/// [`Collect::max_level_hint`] method can also be implemented to allow spans and
+/// events above a maximum verbosity level to be skipped more efficiently,
+/// improving performance.
```

Comment:
```suggestion
/// ## Example
/// Below is a simple example of how a [collector] could implement filtering through
/// a [`LevelFilter`]. When a span or event is recorded, the [`Collect::enabled`] method 
/// compares the span or event's `Level` against the configured [`LevelFilter`].
/// The optional [`Collect::max_level_hint`] method can also be implemented to  allow spans 
/// and events above a maximum verbosity level to be skipped more efficiently,
/// often improving performance in short-lived programs.
```

Reply:
i wasn't sure if this should have an `Example` section here made sense since there were other examples outside of the section. also, this will _always_ improve filtering performance, even if that performance improvement isn't visible over a long-running program's lifespan, so I'm not convinced we need to hedge that.

Reply:
that's reasonable. let's drop the section header. as for short-lived programs: yeah, hedging that might be overkill, but might be helpful to mention that the performance improvement universal, but _most prominent_ in short-lived programs?